File: src/main/java/com/theendercore/hydra/util/Methods.java
package com.theendercore.hydra.util;

import com.theendercore.hydra.config.ModConfig;
import com.theendercore.hydra.mixin.GameRendererAccessor;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.hud.InGameHud;
import net.minecraft.client.render.GameRenderer;
import net.minecraft.text.MutableText;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.random.Random;

import javax.annotation.Nullable;
import java.text.SimpleDateFormat;
import java.util.Date;

import static net.minecraft.util.Formatting.FORMATTING_CODE_PREFIX;
import static com.theendercore.hydra.HydraMod.LOGGER;

public class Methods {
    private static final GameRenderer renderer = MinecraftClient.getInstance().gameRenderer;

    public static void addTwitchMessage(Date date, MutableText usernameText, String message, @Nullable Formatting chatColor, ModConfig c, Boolean isVIP) {
        MutableText timestampText = Text.literal("[" + new SimpleDateFormat(c.getTimeFormatting()).format(date) + "]").formatted(Formatting.GRAY);
        MutableText messageBodyText = Text.literal(": ").formatted(Formatting.WHITE);
        if (!isVIP) {
            message = message.replaceAll(String.valueOf(FORMATTING_CODE_PREFIX), "$");
        }
        if (chatColor == null) {
            messageBodyText.append(Text.literal(message));
        } else {
            messageBodyText.append(Text.literal(message).formatted(chatColor, Formatting.BOLD, Formatting.ITALIC));
        }
        chatMessage(timestampText.append(usernameText).append(messageBodyText));
    }

    public static void chatMessage(Text text) {
        MinecraftClient.getInstance().inGameHud.getChatHud().addMessage(text);
    }

    public static void titleMessage(@Nullable Text text, @Nullable Text smallText) {
        InGameHud hud = MinecraftClient.getInstance().inGameHud;
        if (!(text == null)) {
            hud.setTitle(text);
        }
        if (!(smallText == null)) {
            hud.setSubtitle(smallText);
        }
    }

    public static void setRandomShader() {
        Identifier shader = GameRendererAccessor.getShaderLocations()[Random.create().nextInt(GameRenderer.SHADER_COUNT)];
        LOGGER.info("Loading shader " + shader.getPath());
        ((GameRendererAccessor) renderer).invokeLoadShader(shader);
    }

    public static void disableShader() {
        LOGGER.info("Disabling all shaders");
        renderer.disableShader();
    }

}

File: src/main/kotlin/com/theendercore/hydra/util/Methods.kt
package com.theendercore.hydra.util

import com.theendercore.hydra.config.ModConfig
import net.minecraft.client.MinecraftClient
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.particle.DefaultParticleType
import net.minecraft.particle.ParticleType
import net.minecraft.registry.Registries
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvent
import net.minecraft.text.MutableText
import net.minecraft.text.Text
import net.minecraft.util.Formatting
import java.text.SimpleDateFormat
import java.util.*

object Methods {
    private val client: MinecraftClient = MinecraftClient.getInstance()
    private val renderer = client.gameRenderer
//    private var random: Random = Random.create()
    fun addTwitchMessage(
        date: Date?,
        usernameText: MutableText?,
        msg: String,
        chatColor: Formatting?,
        isVIP: Boolean,
    ) {
        var message = msg
        val timestampText =
            Text.literal("[" + SimpleDateFormat(ModConfig.config!!.timeFormatting).format(date) + "]").formatted(
                Formatting.GRAY
            )
        val messageBodyText = Text.literal(": ").formatted(Formatting.WHITE)

        if (!isVIP) message = message.replace(Formatting.FORMATTING_CODE_PREFIX.toString().toRegex(), "$")

        if (chatColor == null) messageBodyText.append(Text.literal(message))
        else messageBodyText.append(Text.literal(message).formatted(chatColor, Formatting.BOLD, Formatting.ITALIC))

        chatMessage(timestampText.append(usernameText).append(messageBodyText))
    }

    fun chatMessage(text: Text?) {
        client.inGameHud.chatHud.addMessage(text)
    }

    fun titleMessage(text: Text?, smallText: Text?) {
        val hud = client.inGameHud
        if (text != null) {
            hud.setTitle(text)
        } else {
            hud.setSubtitle(smallText)
        }
    }

    fun setRandomShader() {
//        renderer.cycleSuperSecretSetting()
    }

    fun disableShader() {
//        client.gameRenderer.disablePostProcessor()
    }

    fun playSound(player: PlayerEntity, sound: SoundEvent) {
        player.playSound(sound, SoundCategory.PLAYERS, 1f, 1f)
    }

    fun playRandomSound(player: PlayerEntity) {
//        Registries.SOUND_EVENT[random.nextBetween(0, Registries.SOUND_EVENT.size() - 1)]?.let { playSound(player, it) }
    }

    fun playParticle(player: PlayerEntity, particle: ParticleType<*>): Int {
//        for (i in 0 until 100) {
//            try {
//                client.particleManager.addParticle(
//                    particle as ParticleEffect,
//                    player.x + random.nextGaussian() * 1f,
//                    player.y + 1.8 + random.nextGaussian() * 1f,
//                    player.z + random.nextGaussian() * 1f,
//                    random.nextGaussian() * 1f,
//                    random.nextGaussian() * 1f,
//                    random.nextGaussian() * 1f
//                )
//            } catch (throwable: Throwable) {
//                LOGGER.warn(
//                    "Could not spawn particle effect {}",
//                    if (particle is DefaultParticleType) particle.asString() else particle
//                )
//                return 0
//            }
//        }
        return 1
    }

    fun randomParticle(player: PlayerEntity): Int {
        val particles = Registries.PARTICLE_TYPE.filterIsInstance<DefaultParticleType>()
//        return playParticle(player, particles[random.nextBetween(0, particles.size - 1)])
        return 0
    }
}